kernel/mem/virtmem.cpp

```diff
@@ -1,34 +1,134 @@
 /**
  * 
  */
+#include <kernel/device/uart.h>
+#include <kernel/kstdlib/kstdio.h>
+#include <kernel/mem/address.h>
 #include <kernel/mem/virtmem.h>
 
-namespace VirtMemoryManager
-{
+extern uint32_t __RAM_START;
+extern uint32_t __RAM_END;
+extern uint32_t __FREE_RAM_START;
+static constexpr uint32_t PHYS_PAGE_SIZE = 0x1000;
+static constexpr uint32_t L1_TABLE_ENTRY_DEFAULT = 0x00000001;
+static constexpr uint32_t L2_TABLE_ENTRY_DEFAULT = 0x00000032;
+static constexpr uint32_t DEVICES_MEMORY_BASE = 0x1c000000;
+static constexpr uint32_t GIC_MEMORY_BASE = 0x2c000000;
 
-static uint16_t L1_translation_table_index(VirtualAddress address)
+VirtMemoryManager::VirtMemoryManager()
 {
-    return address.get() >> 20;
+    /* First we build the translation table to simply map the kernel and devices */
+    L1TranslationTable* l1_table_base = allocate_L1_translation_table();
+    initialise_L1_translation_table(*l1_table_base);
+    identity_map_kernel(*l1_table_base);
+    map_devices(*l1_table_base);
+
+    /* Now we can use this table as our map and enable the MMU */
+    switch_L1_translation_table(l1_table_base);
+    CPU::invalidate_unified_TLB();
+    CPU::set_DACR_by_domain(Domain::ALL_DOMAINS, DACRValue::CLIENT);
+    CPU::set_SCTLR_flag(SCTLRFlag::MMU_ENABLE);
```

Comment:
We lose these tables completely when this function returns? Shouldn't the Kernel's Page Directory/L1Table be something that's a data member of the virtmem class?